csrc/dynamic_transform.cpp

```diff
@@ -943,6 +946,90 @@ void DynamicTransformConcretizer::mutate(TensorDomain* td) {
   registerConcretization(td, mutated_val);
 }
 
+// Maybe insert SqueezeOps on inputs of ReductionOp, to simplify trivial
+// reductions.
+void DynamicTransformConcretizer::mutate(Expr* expr) {
+  const auto mutateAndReplaceInOutputs = [this](Val* old_val, Val* new_val) {
+    registerMutation(old_val, new_val);
+    if (old_val->isFusionOutput()) {
+      old_val->fusion()->replaceOutput(old_val, new_val);
```

Comment:
I'm not sure if this is a real concern, but we are delaying the mutation after finishing the traversal is done since it's really error prone to modify a fusion while traversing it. Since `replaceOutput` is also a modification, shouldn't it also be delayed?

Reply:
Great question. This pass actually already modifies the fusion by rewriting `Expr`s; that's what `mutate(Expr*)` does. However, there's a little bit of a subtlety here because if we change outputs, that can actually change `uses()` during traversal. I don't think that's actually important in this case, but in order to avoid tricky bugs, I think I'll add a stage at the end of concretization where we loop over outputs and replace them if they're registered for mutation.